ksql-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTarget.java

```diff
@@ -171,89 +178,123 @@ private KsqlRequest ksqlRequest(
-      final Optional<Integer> readTimeoutMs
+      final Object requestBody,
+      final Function<ResponseWithBody, T> mapper
   ) {
-    try {
-      // Performs an asynchronous request
-      return target
-          .path(path)
-          .request(MediaType.APPLICATION_JSON_TYPE)
-          .property(ClientProperties.READ_TIMEOUT, readTimeoutMs.orElse(0))
-          .headers(headers())
-          .async()
-          .post(Entity.json(jsonEntity));
-    } catch (final ProcessingException e) {
-      if (shouldRetry(readTimeoutMs, e)) {
-        return postAsync(path, jsonEntity, calcReadTimeout(readTimeoutMs));
-      }
-      throw new KsqlRestClientException("Error issuing POST to KSQL server. path:" + path, e);
-    } catch (final Exception e) {
-      throw new KsqlRestClientException("Error issuing POST to KSQL server. path:" + path, e);
-    }
+    return executeSync(httpMethod, path, requestBody, mapper, (resp, vcf) -> {
+      resp.bodyHandler(buff -> vcf.complete(new ResponseWithBody(resp, buff)));
+    });
   }
 
-  private MultivaluedMap<String, Object> headers() {
-    final MultivaluedMap<String, Object> headers = new MultivaluedHashMap<>();
-    authHeader.ifPresent(v -> headers.add(HttpHeaders.AUTHORIZATION, v));
-    return headers;
+  private <T> RestResponse<StreamPublisher<T>> executeQueryRequestWithStreamResponse(
+      final String ksql,
+      final Optional<Long> previousCommandSeqNum,
+      final Function<Buffer, T> mapper
+  ) {
+    final KsqlRequest ksqlRequest = createKsqlRequest(ksql, previousCommandSeqNum);
+    final AtomicReference<StreamPublisher<T>> pubRef = new AtomicReference<>();
+    return executeSync(HttpMethod.POST, QUERY_PATH, ksqlRequest, resp -> pubRef.get(),
+        (resp, vcf) -> {
+          if (resp.statusCode() == 200) {
+            pubRef.set(new StreamPublisher<>(Vertx.currentContext(),
+                resp, mapper, vcf));
+            vcf.complete(new ResponseWithBody(resp));
+          } else {
+            resp.bodyHandler(body -> vcf.complete(new ResponseWithBody(resp, body)));
+          }
+        });
   }
 
-  private static boolean shouldRetry(
-      final Optional<Integer> readTimeoutMs,
-      final ProcessingException e
+  private <T> RestResponse<T> executeSync(
+      final HttpMethod httpMethod,
+      final String path,
+      final Object requestBody,
+      final Function<ResponseWithBody, T> mapper,
+      final BiConsumer<HttpClientResponse, CompletableFuture<ResponseWithBody>> responseHandler
   ) {
-    return readTimeoutMs.map(timeout -> timeout < MAX_TIMEOUT).orElse(false)
-        && e.getCause() instanceof SocketTimeoutException;
+    final CompletableFuture<ResponseWithBody> vcf =
+        execute(httpMethod, path, requestBody, responseHandler);
+
+    final ResponseWithBody response;
+    try {
+      response = vcf.get();
+    } catch (Exception e) {
+      throw new KsqlRestClientException("Error issuing GET to KSQL server. path:" + path, e);
```

Comment:
`GET` hardcoded for httpMethod in error message.

Reply:
ack